src/backend/commands/vacuum.c

```diff
@@ -2687,23 +2687,33 @@ vacuum_rel(Oid relid, RangeVar *relation, VacuumParams *params,
 	 * Don't dispatch auto-vacuum. Each segment performs auto-vacuum as per
 	 * its own need.
 	 */
-	if (Gp_role == GP_ROLE_DISPATCH && !recursing &&
+	if ((Gp_role == GP_ROLE_DISPATCH || Gp_role == GP_ROLE_SINGLENODE) && !recursing &&
```

Comment:
why singlenode need this logical?

Reply:
As I commented below, from my understanding, it is unneccesary to do that. But there are some more checks below and suggests that we should already in some transaction. It's a little knoty to solve it, and since this pr is already large enough, let's fix some issue like this in the future.